#! /usr/bin/env python

PACKAGE="visensor_node"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

# ========================= strobe ==========================

gen.add("strobe",       int_t,    0,    "strobe", 0,     0, 12287)

gen.add("strobe_mode",       bool_t,    0,    "flashing led", True)

# ========================= CAMERA 0 ==========================
#       Name       Type      Level Description     Default Min   Max
gen.add("cam0_name", str_t,    0,    "camera 0", "cam0")
gen.add("cam0_agc_enable",       bool_t,    0,    "automatic gain control", True)
gen.add("cam0_max_analog_gain",       int_t,    0,    "max analog gain", 64,     16, 64)
gen.add("cam0_global_analog_gain",       int_t,    0,    "analog gain", 16,     16, 64)
gen.add("cam0_global_analog_gain_attenuation",       bool_t,    0,    "gain attenuation", False)

gen.add("digital_gain",       int_t,    0,    "digital gain", 4,     1, 15)


gen.add("cam0_aec_enable",       bool_t,    0,    "automatic exposure control", True)
gen.add("cam0_min_coarse_shutter_width",       int_t,    0,    "min shutter width", 2,     2, 32765)
gen.add("cam0_max_coarse_shutter_width",       int_t,    0,    "max shutter width", 500,     2, 32765)
gen.add("cam0_coarse_shutter_width",       int_t,    0,    "shutter", 480,     2, 32765)
gen.add("cam0_fine_shutter_width",       int_t,    0,    "shutter", 0,     0, 1774)

gen.add("cam0_adc_mode",       int_t,    0,    "ADC Mode", 2,2,3)
gen.add("cam0_vref_adc_voltage_level",       int_t,    0,    "Vref ADC Voltage level", 4,0,7)

gen.add("cam0_black_level_calibration_manual_override",       bool_t,    0,    "Black level calibration manual override", False)
gen.add("cam0_black_level_calibration_value",       int_t,    0,    "Black level calibration value", 0,-127,127)


gen.add("cam0_row_flip",       bool_t,    0,    "row flip", False)
gen.add("cam0_column_flip",       bool_t,    0,    "column flip", False)

# ========================= CAMERA 1 ==========================

gen.add("cam1_name", str_t,    0,    "camera 1", "cam1")
gen.add("cam1_agc_enable",       bool_t,    0,    "automatic gain control", True)
gen.add("cam1_max_analog_gain",       int_t,    0,    "max analog gain", 64,     16, 64)
gen.add("cam1_global_analog_gain",       int_t,    0,    "analog gain", 16,     16, 64)
gen.add("cam1_global_analog_gain_attenuation",       bool_t,    0,    "gain attenuation", False)

gen.add("cam1_aec_enable",       bool_t,    0,    "automatic exposure control", True)
gen.add("cam1_min_coarse_shutter_width",       int_t,    0,    "min shutter width", 2,     2, 32765)
gen.add("cam1_max_coarse_shutter_width",       int_t,    0,    "max shutter width", 500,     2, 32765)
gen.add("cam1_coarse_shutter_width",       int_t,    0,    "shutter", 480,     2, 32765)

gen.add("cam1_adc_mode",       int_t,    0,    "ADC Mode", 2,2,3)
gen.add("cam1_vref_adc_voltage_level",       int_t,    0,    "Vref ADC Voltage level", 4,0,7)

gen.add("cam1_row_flip",       bool_t,    0,    "row flip", True)
gen.add("cam1_column_flip",       bool_t,    0,    "column flip", True)

# ========================= CAMERA 2 ==========================

gen.add("cam2_name", str_t,    0,    "camera 2", "cam2")
gen.add("cam2_agc_enable",       bool_t,    0,    "automatic gain control", True)
gen.add("cam2_max_analog_gain",       int_t,    0,    "max analog gain", 64,     16, 64)
gen.add("cam2_global_analog_gain",       int_t,    0,    "analog gain", 16,     16, 64)
gen.add("cam2_global_analog_gain_attenuation",       bool_t,    0,    "gain attenuation", False)

gen.add("cam2_aec_enable",       bool_t,    0,    "automatic exposure control", True)
gen.add("cam2_min_coarse_shutter_width",       int_t,    0,    "min shutter width", 2,     2, 32765)
gen.add("cam2_max_coarse_shutter_width",       int_t,    0,    "max shutter width", 500,     2, 32765)
gen.add("cam2_coarse_shutter_width",       int_t,    0,    "shutter", 480,     2, 32765)

gen.add("cam2_adc_mode",       int_t,    0,    "ADC Mode", 2,2,3)
gen.add("cam2_vref_adc_voltage_level",       int_t,    0,    "Vref ADC Voltage level", 4,0,7)

gen.add("cam2_row_flip",       bool_t,    0,    "row flip", True)
gen.add("cam2_column_flip",       bool_t,    0,    "column flip", True)

# ========================= CAMERA 3 ==========================

gen.add("cam3_name", str_t,    0,    "camera 3", "cam3")
gen.add("cam3_agc_enable",       bool_t,    0,    "automatic gain control", True)
gen.add("cam3_max_analog_gain",       int_t,    0,    "max analog gain", 64,     16, 64)
gen.add("cam3_global_analog_gain",       int_t,    0,    "analog gain", 16,     16, 64)
gen.add("cam3_global_analog_gain_attenuation",       bool_t,    0,    "gain attenuation", False)

gen.add("cam3_aec_enable",       bool_t,    0,    "automatic exposure control", True)
gen.add("cam3_min_coarse_shutter_width",       int_t,    0,    "min shutter width", 2,     2, 32765)
gen.add("cam3_max_coarse_shutter_width",       int_t,    0,    "max shutter width", 500,     2, 32765)
gen.add("cam3_coarse_shutter_width",       int_t,    0,    "shutter", 100,     2, 32765)

gen.add("cam3_adc_mode",       int_t,    0,    "ADC Mode", 2,2,3)
gen.add("cam3_vref_adc_voltage_level",       int_t,    0,    "Vref ADC Voltage level", 4,0,7)

gen.add("cam3_row_flip",       bool_t,    0,    "row flip", True)
gen.add("cam3_column_flip",       bool_t,    0,    "column flip", True)

# ========================= Dense Matcher ==========================
gen.add("penalty_1",       int_t,    0,    " ", 2,     0, 32000)
gen.add("penalty_2",       int_t,    0,    " ", 60,     0, 32000)
gen.add("threshold",       int_t,    0,    " ", 255,     0, 32000)
gen.add("lr_check",       int_t,    0,    " ", 5,     0, 32000)

# ========================= External Trigger ==========================
gen.add("enable_trigger_0",       bool_t,    0,    "Enable first trigger line", False)
gen.add("enable_trigger_1",       bool_t,    0,    "Enable sedond trigger line", False)
gen.add("enable_trigger_2",       bool_t,    0,    "Enable third trigger line", False)

direction_enum = gen.enum([ gen.const("Output",      bool_t, False, "Set trigger direction to output"),
	gen.const("Input",     bool_t, True, "Set trigger direction to input")], "An enum to choose direction")


gen.add("trigger_0_direction",       bool_t,    0,    "Sets trigger direction.", True, edit_method=direction_enum)
gen.add("trigger_1_direction",       bool_t,    0,    "Sets trigger direction.", True, edit_method=direction_enum)
gen.add("trigger_2_direction",       bool_t,    0,    "Sets trigger direction.", True, edit_method=direction_enum)

gen.add("trigger_0_rate",       int_t,    0,    "Rate of trigger line in Hz ", 0, 0, 1000)
gen.add("trigger_1_rate",       int_t,    0,    "Rate of trigger line in Hz ", 0, 0, 1000)
gen.add("trigger_2_rate",       int_t,    0,    "Rate of trigger line in Hz ", 0, 0, 1000)

gen.add("trigger_0_offset",       int_t,    0,    "Offset of trigger line in 10 microseconds steps", 0, 0, 100000)
gen.add("trigger_1_offset",       int_t,    0,    "Offset of trigger line in 10 microseconds steps", 0, 0, 100000)
gen.add("trigger_2_offset",       int_t,    0,    "Offset of trigger line in 10 microseconds steps", 0, 0, 100000)

exit(gen.generate(PACKAGE, "visensor_node", "Driver"))

